README.md
# auzy

[![Build Status](https://travis-ci.org/alexey-detr/auzy.svg?branch=master)](https://travis-ci.org/alexey-detr/auzy) [![Maintainability](https://api.codeclimate.com/v1/badges/dc7769e214a244cb68aa/maintainability)](https://codeclimate.com/github/alexey-detr/auzy/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/dc7769e214a244cb68aa/test_coverage)](https://codeclimate.com/github/alexey-detr/auzy/test_coverage)

Don't ever get confused when adding session middleware into your project!

## Project features

- Modern ES6 syntax
- Promise and async/await ready
- Compatible with many middleware based frameworks (e.g. Express, Restify)
- Flexible to use any session storage (e.g. Redis, MongoDB)
- Well documented (JSDoc is in progress)
- Made as simple as possible
- Covered with unit tests
- Covered with functional tests for following frameworks:
  - [Restify](http://restify.com/)
  - [Express](http://expressjs.com/)
  - [koa.js](http://koajs.com/)
  - [Connect](https://github.com/senchalabs/connect)

It's all about **auzy**. The Node.js middleware library made to add sessions support.

**The project is ready for use, but it is not tested well yet, so please be careful using it.**

## Quick start

_For full usage examples with specific framework see the [examples directory](https://github.com/alexey-detr/auzy/tree/master/examples)._

Install auzy and session storage (Redis in this example)

```bash
npm install --save auzy auzy-storage-redis
# or
yarn add auzy auzy-storage-redis
```

Add the middleware into your project

```js
const auzyConfig = {
    session: {
        // Session name will be used as header name in this example.
        sessionName: 'X-Session-Token',

        // TTL is a session expiration time in milliseconds.
        ttl: 60 * 60 * 24 * 30 * 6 * 1000,

        // Contains a logic to fetch user data from database.
        // Promise is expected as return value, it has to resolve user data.
        // Resolved user data will be set as req.user for current request.
        // Or as ctx.state.user in case of Koa framework.
        loadUser: (sessionData) => {
            const query = {_id: new ObjectId(sessionData.userId)};
            return User.Model.findOne().exec();
        },
    },
};
const auzyEnvironment = {
    storage: 'redis',
    framework: 'express',
    transport: 'header',
};
app.use(auzy(auzyConfig, auzyEnvironment));
```

### Login

Use session object inside your login route

```js
async (req, res, next) => {
    const reqUser = req.body.user;
    const dbQuery = {
        email: reqUser.email,
    };
    const user = await User.Model.findOne(dbQuery).exec();
    if (!user) {
        return next();
    }
    await user.validatePassword(reqUser.password);
    await req.session.authenticate({userId: String(user._id)});
    res.send({user: req.user.name});
}
```

### Logout

To logout and destroy current session you have to call `destroy()` method

```js
async (req, res, next) => {
    const user = req.user;
    if (user) {
        // It is important to wait the destroying process
        // so after that line we will be sure that there is nothing
        // left from a previously authorized user.
        await req.session.destroy();
    }
    // User is already deleted from the request object
    // so we can safely use it in a general way, it will
    // be undefined here.
    res.send({user: req.user});
}
```

## Configuration

There are two things you can specify in the configuration:

- Session handler configuration
- Environment hints

### Session handler

Session handler is a core of auzy project, which has configuration options described in this section

##### `function receiveSessionId(adapter, sessionName): string`

The method describes how the sessionId will be fetched from the request.

##### `function sendSessionId(adapter, sessionName, sessionId)`

The method describes how the sessionId will be put into the response.

##### `sessionName: string`

Property specifies the session name, which will be passed to `receiveSessionId` or `sendSessionId` methods.

##### `function loadUser(sessionData): Promise`

The method describes how user will be loaded using `sessionData`. User object itself can be stored into your main application database, e.g. MongoDB. The returned promise should resolve user object fetched from DB. Fetched user will be stored in the request object (or something similar so you will be able to access it in any controller).

##### `function generateSessionId(): string`

The method describes how the new session ID will be generated. If not specified, the session ID will be Base64 encoded 20 bytes length cryptographically strong pseudo-random data. To generate it auzy uses node's module [crypto](https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback).

##### `ttl: string`

Property specifies how long the session data will be valid until expiration. **Value must be specified in milliseconds.**

##### `alwaysSave: boolean`

If `true` the session data will be saved every time the session is loaded. Can be useful to resave the new user name if it is stored in the session data.

##### `alwaysSend: boolean`

If `true` the session ID will be send to client in the every response. If `false` the session ID will be send only when `session.authenticate()` is explicitly called.

### Environment hints

Environment hints help auzy understand which kind of storage it should use or which framework context it is going to work in.
So hint can describe framework or storage.

##### `framework: string`

Just a framework code. Following frameworks are supported for now:

- `restify` — [Restify](http://restify.com/)
- `express` — [Express](http://expressjs.com/)
- `koa` — [koa.js](http://koajs.com/)
- `connect` — [Connect](https://github.com/senchalabs/connect)

If the framework hint is specified, `receiveSessionId` and `sendSessionId` will get in their first argument adapter object, which actually just a unified way to modify request or response object for hinted framework.

##### `storage: string`

Storage hint specifies which kind of storage will be used to store session data. Following storages are supported now:

- `redis` — Redis [alexey-detr/auzy-storage-redis](https://github.com/alexey-detr/auzy-storage-redis)

If not specified a native JS object storage will be used. It is non-persistent so any saved session data will be destroyed whenever the server will be restarted.

##### `transport: string`

Transport hint specifies how session ID will be transferred to client or from client. `header` is the only available transport for now. It passes session ID in the HTTP header named as `session.sessionName` (in the example above it was `X-Session-Token`).

## Advanced config example

If your framework is not supported you can try to configure it with more complex config.

```js
const auzyConfig = {
    session: {
        // Session name will appear as request and response header name below.
        sessionName: 'X-Session-Token',

        // TTL is the session expiration time in milliseconds.
        ttl: 60 * 60 * 24 * 30 * 6 * 1000,

        // Contains logic to fetch session ID from request object.
        // String with session ID must be returned.
        receiveSessionId: (req, sessionName) => req.header(sessionName),

        // Contains logic to send session ID to client via response object.
        sendSessionId: (res, sessionName, sessionId) => res.header(sessionName, sessionId),

        // Contains logic to fetch user data from database.
        // Promise is expected as return value, it should resolve user data.
        // Resolved user data will be set as req.user for current request.
        loadUser: (sessionData) => {
            const query = {_id: new ObjectId(sessionData.userId)};
            return User.Model.findOne().exec();
        }
    },
};
const auzyEnvironment = {storage: 'redis'};
app.use(auzy(auzyConfig, auzyEnvironment));
```

## TODOs and ideas

- [ ] Allow to store any data in the session, not just authorization data.
- [ ] Various transports or ways to get / pass session IDs from / to client (header, cookie, body?)
- [ ] MongoDB session storage
- [ ] Authenticators support to validate credentials (e.g. email password pairs, any kind of tokens) with hashes or other verifiers. In other words it will provide easy to use Facebook login for example.
- [ ] Made project ready to support things like JWT which use client as a session storage.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="4.0.0"></a>
# [4.0.0](https://github.com/alexey-detr/auzy/compare/v3.0.0...v4.0.0) (2018-06-11)


### Bug Fixes

* Fixed authenticate() method to always send sessionId ([620edcb](https://github.com/alexey-detr/auzy/commit/620edcb))


### Features

* Session ID will be regenerated on every authenticate() call ([00d7a60](https://github.com/alexey-detr/auzy/commit/00d7a60))


### BREAKING CHANGES

* Calling of authenticate() method will now regenerate a session ID



<a name="3.0.0"></a>
# [3.0.0](https://github.com/alexey-detr/auzy/compare/v2.1.0...v3.0.0) (2017-12-18)


### Bug Fixes

* fixed issue when user was not deleted from session in Koa ([ed1117e](https://github.com/alexey-detr/auzy/commit/ed1117e))


### Features

* **ObjectStorage:** Added TTL support for a native ObjectStorage ([44da244](https://github.com/alexey-detr/auzy/commit/44da244))
* added header transport implementation ([d93d5d7](https://github.com/alexey-detr/auzy/commit/d93d5d7))


### BREAKING CHANGES

* **ObjectStorage:** TTL has to be specified in milliseconds now



<a name="2.1.0"></a>
# [2.1.0](https://github.com/alexey-detr/auzy/compare/v2.0.0...v2.1.0) (2017-12-08)


### Features

* Added first adapters implementation ([29439a0](https://github.com/alexey-detr/auzy/commit/29439a0))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/alexey-detr/auzy/compare/v2.0.0...v2.0.1) (2017-12-08)


### Features

* Added first adapters implementation ([29439a0](https://github.com/alexey-detr/auzy/commit/29439a0))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/alexey-detr/auzy/compare/v1.0.2...v2.0.0) (2017-12-03)


### Tests

* First functional tests for Restify framework ([8d6c778](https://github.com/alexey-detr/auzy/commit/8d6c778))


### BREAKING CHANGES

* Middleware's second parameter is an object now, it contains definition of storage,
framework, etc.



<a name="1.0.2"></a>
## [1.0.2](https://github.com/alexey-detr/auzy/compare/v1.0.1...v1.0.2) (2017-11-20)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/alexey-detr/auzy/compare/e90b869...v1.0.1) (2017-11-19)


### Code Refactoring

* Now req and res are captured for every request ([c229dba](https://github.com/alexey-detr/auzy/commit/c229dba))


### Features

* Now it is possible to specify receiveSessionId in config ([e90b869](https://github.com/alexey-detr/auzy/commit/e90b869))


### BREAKING CHANGES

* API was broken


examples/connect.js
const connect = require('connect');
const bodyParser = require('body-parser');
const auzy = require('../index');

const users = [{
    id: 1,
    name: 'Bob',
    email: 'bob@mail.com',
}];

const auzyConfig = {
    session: {
        sessionName: 'X-Session-Token',
        ttl: 60 * 60 * 24 * 30 * 6,
        alwaysSend: true,
        loadUser: (sessionData) => {
            const index = users.findIndex(user => user.id === sessionData.userId);
            return users[index];
        },
    },
};
const auzyEnvironment = {
    framework: 'connect',
    transport: 'header',
};

const server = connect();
server.use(bodyParser.json());
server.use(auzy(auzyConfig, auzyEnvironment));

server.use('/login', async (req, res) => {
    const index = users.findIndex(user => user.name === req.body.name);
    if (index !== -1) {
        const user = users[index];
        await req.session.authenticate({userId: user.id});
        res.end(JSON.stringify({name: req.user.name}));
    }
});

server.use('/secret', (req, res) => {
    if (req.user) {
        res.end(JSON.stringify({email: req.user.email}));
    } else {
        res.statusCode = 403;
        res.end(JSON.stringify({error: 'Restricted area'}));
    }
});

server.use('/logout', async (req, res) => {
    await req.session.destroy();
    // send the user object back just to be sure that the user is not accessible
    // after the session was destroyed
    res.end(JSON.stringify({user: req.user}));
});

const launchPromise = new Promise((resolve) => {
    const nodeServer = server.listen(9001, () => {
        resolve(nodeServer);
    });
});

module.exports = launchPromise;
